modules/apps/oauth2-provider/oauth2-provider-rest/src/main/java/com/liferay/oauth2/provider/rest/internal/endpoint/access/token/grant/handler/LiferayResourceOwnerAccessTokenGrantHandler.java

```diff
@@ -52,15 +52,15 @@
 	protected void activate(Map<String, Object> properties) {
 		_resourceOwnerGrantHandler = new ResourceOwnerGrantHandler();
 
-		_resourceOwnerGrantHandler.setDataProvider(_liferayOAuthDataProvider);
-		_resourceOwnerGrantHandler.setLoginHandler(_resourceOwnerLoginHandler);
-
 		_oAuth2ProviderConfiguration = ConfigurableUtil.createConfigurable(
 			OAuth2ProviderConfiguration.class, properties);
 	}
 
 	@Override
 	protected AccessTokenGrantHandler getAccessTokenGrantHandler() {
+		_resourceOwnerGrantHandler.setDataProvider(_liferayOAuthDataProvider);
+		_resourceOwnerGrantHandler.setLoginHandler(_resourceOwnerLoginHandler);
```

Comment:
@W2jx It is really weird that you have a local variable, and each time you retrieve it, you set it. Usually for those you initialize once and reuse later, we make it a local variable, here it should not be a local variable.

Also it is really weird that `LiferayAuthorizationAccessTokenCodeGrantHandler` is a `AccessTokenGrantHandler`, and it has a `getAccessTokenGrantHandler` method which did not return itself, but another type of `AccessTokenGrantHandler`.

You should not have `getAccessTokenGrantHandler()`, instead, let `LiferayAuthorizationAccessTokenCodeGrantHandler` directly use `AuthorizationCodeGrantHandler` to implement the logic.

FYI @dantewang